R/calculate_tdm.R

```diff
@@ -0,0 +1,305 @@
+  data %>%
+    mutate(
+      time_step = case_when(
+        .data$year == t0 ~ "t0",
+        .data$year == t0 + t1 ~ "plus_t1",
+        .data$year == t0 + t2 ~ "plus_t2"
+      )
+    )
+}
+
+add_monotonic_factor <- function(data, t0, t1, t2, groups) {
+  data <- data %>%
+    group_by(!!!rlang::syms(groups)) %>%
+    mutate(
+      end_year = last(.data$year),
+      end_year_is_t0_t2 = .data$end_year == t0 + t2
+    )
+
+  if (all(data$end_year_is_t0_t2)) {
+    monotonic_factors <- data %>%
+      mutate(
+        .is_monotonic = TRUE,
+        monotonic_factor = dplyr::if_else(.data$.is_monotonic, 1, -1),
+        .is_monotonic = NULL
+      ) %>%
+      select(
+        -c(
+          .data$year,
+          .data$end_year,
+          .data$plan_carsten,
+          .data$plan_alloc_wt_tech_prod,
+          .data$scen_alloc_wt_tech_prod,
+          .data$end_year_is_t0_t2
+        )
+      ) %>%
+      distinct() %>%
+      ungroup()
+  } else {
+    monotonic_factors <- data %>%
+      add_time_step(t0, t1, t2) %>%
+      mutate(time_step = case_when(
+        .data$year == .data$end_year ~ "end_year",
+        TRUE ~ .data$time_step
+      )) %>%
+      select(
+        -c(
+          .data$year,
+          .data$end_year,
+          .data$plan_carsten,
+          .data$plan_alloc_wt_tech_prod,
+          .data$end_year_is_t0_t2
+        )
+      ) %>%
+      tidyr::pivot_wider(
+        names_from = .data$time_step,
+        values_from = .data$scen_alloc_wt_tech_prod
+      ) %>%
+      mutate(
+        .increasing_overall = .data$end_year > .data$t0,
+        .increasing_in_interval = .data$plus_t2 > .data$t0,
+        .is_monotonic = .data$.increasing_in_interval & .data$.increasing_overall,
+        monotonic_factor = dplyr::if_else(.data$.is_monotonic, 1, -1),
+        .increasing_overall = NULL,
+        .increasing_in_interval = NULL,
+        .is_monotonic = NULL,
+        t0 = NULL,
+        plus_t1 = NULL,
+        plus_t2 = NULL,
+        end_year = NULL
+      ) %>%
+      ungroup()
+  }
+
+  left_join(data, monotonic_factors, by = groups)
+}
+
+add_tdm <- function(data, groups) {
+  data %>%
+    group_by(!!!rlang::syms(groups)) %>%
+    mutate(
+      .numerator = .data$scenario_production_plus_t2 - .data$portfolio_production_plus_t1,
+      .denominator = .data$scenario_production_plus_t2 - .data$scenario_production_t0,
+      # TODO: This case we
+      tdm_tech = ifelse(
+        .data$.denominator == 0,
+        0,
+        max(0, (.data$.numerator / .data$.denominator) * .data$monotonic_factor) * 2
+      ),
+      .numerator = NULL,
+      .denominator = NULL
+    ) %>%
+    select(.data$tdm_tech, all_of(groups)) %>%
+    distinct() %>%
+    ungroup()
+}
+
+add_aggregate_tdm <- function(data, t0, groups) {
+  data %>%
+    group_by(!!!rlang::syms(groups)) %>%
+    ungroup(.data$technology) %>%
+    mutate(
+      tdm_sec = .data$plan_carsten * sum(.data$tdm_tech) / sum(.data$plan_carsten),
```

Comment:
do you need to use `na.rm = FALSE` here to be safe?

Reply:
`na.rm = FALSE` is the default, no?

Reply:
sorry, I meant `na.rm = TRUE`.... if there's any chance at all that `.data$tdm_tech` or `.data$plan_carsten` could have an `NA` in them (or a `NULL`), you definitely want to use `na.rm = TRUE`, because otherwise this will resolve to  `NA`, which is probably not what you want (the default sucks imo), e.g.
```r
sum(c(1, 2, 3))
# [1] 6
sum(c(1, 2, 3, NA))
# [1] NA
sum(c(1, 2, 3, NA), na.rm = TRUE)
# [1] 6
```